### PR TITLE
Add table oci_apigateway_api. closes #127

### LINKED ISSUE #127: Add table oci_apigateway_api
**References**
Add any related links that will help us understand the resource, including vendor documentation, related GitHub issues, and Go SDK documentation.

### PR DESCRIPTION
**Pull request by Priyanka585464**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
SETUP: tests/oci_apigateway_api []

PRETEST: tests/oci_apigateway_api

TEST: tests/oci_apigateway_api
Running terraform

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # oci_apigateway_api.named_test_resource will be created
  + resource "oci_apigateway_api" "named_test_resource" {
      + compartment_id     = "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq"
      + content            = (known after apply)
      + defined_tags       = (known after apply)
      + display_name       = "steampipetest7265"
      + freeform_tags      = (known after apply)
      + id                 = (known after apply)
      + lifecycle_details  = (known after apply)
      + specification_type = (known after apply)
      + state              = (known after apply)
      + time_created       = (known after apply)
      + time_updated       = (known after apply)
      + validation_results = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + reporting_region = "ap-mumbai-1"
  + resource_id      = (known after apply)
  + resource_name    = "steampipetest7265"
  + tenancy_ocid     = "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq"
oci_apigateway_api.named_test_resource: Creating...
oci_apigateway_api.named_test_resource: Creation complete after 1s [id=ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaahoatiifd3usakadcmmrjzzwidsdqbh3elnfdjjb6l7dq]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

reporting_region = "ap-mumbai-1"
resource_id = "ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaahoatiifd3usakadcmmrjzzwidsdqbh3elnfdjjb6l7dq"
resource_name = "steampipetest7265"
tenancy_ocid = "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq"

Running SQL query: test-get-query.sql

[
  {
    "display_name": "steampipetest7265",
    "id": "ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaahoatiifd3usakadcmmrjzzwidsdqbh3elnfdjjb6l7dq"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql

[
  {
    "display_name": "steampipetest7265",
    "id": "ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaahoatiifd3usakadcmmrjzzwidsdqbh3elnfdjjb6l7dq"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql

null
✔ PASSED

Running SQL query: test-turbot-query.sql

[
  {
    "tenant_id": "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq",
    "title": "steampipetest7265"
  }
]
✔ PASSED

POSTTEST: tests/oci_apigateway_api

TEARDOWN: tests/oci_apigateway_api

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

```sql
select
  id,
  display_name,
  time_created,
  lifecycle_state as state
from
  oci_apigateway_api;
```
```
+--------------------------------------------------------------------------------------------------+-----------------------------+---------------------+---------+
| id                                                                                               | display_name                | time_created        | state   |
+--------------------------------------------------------------------------------------------------+-----------------------------+---------------------+---------+
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaag5mrulvgredt2sixmom52bjkz7agqsckzejjstz67awa | turbotsteampipe             | 2021-05-27 13:15:59 | DELETED |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaa7gfxiddk37jlfgsq4xwvomekwcqssvfip5amqsf2ronq | steampipetest1625           | 2021-05-31 12:48:04 | DELETED |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaahoatiifd3usakadcmmrjzzwidsdqbh3elnfdjjb6l7dq | steampipetest7265           | 2021-05-31 12:48:41 | DELETED |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaaitbtzu76ilnaatrkayscncje522kqyboo4ztg4swhhca | turbotsteampipe             | 2021-05-27 13:32:40 | ACTIVE  |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaaaakr7cvrpyht2cit2wngoulfijactl7vx7xi4uob4hla | apigatewayapi20210531121226 | 2021-05-31 12:12:26 | DELETED |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaa746v4gi6ty5zrdvmvv3h7t72ev7daf346ccih6krkvja | steampipetest5615           | 2021-05-31 12:44:36 | DELETED |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaarzd76rlohddfi4b2rvyjrhilpdqgz6j2zkovc2bh7jia | turbotsteampipe             | 2021-05-27 13:22:37 | ACTIVE  |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaapu6hg2nc2oykwqpybwqz76kuqttvw2sjii2mkj2eag5q | apigatewayapi20210531091235 | 2021-05-31 09:12:35 | DELETED |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaakauyn3llywz3netgt765rmcvf6q65cskjyzetq5geqla | apigatewayapi20210531121023 | 2021-05-31 12:10:23 | DELETED |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaaz3fkv6ghzrr6qwfx6mh436fhhmacuizmbzjzr5nukj3a | apigatewayapi20210531123414 | 2021-05-31 12:34:14 | DELETED |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaa3sfgzvahufhqefercu2dpbaybqcimpmr7etpdtwjpwqa | apigatewayapi20210531121959 | 2021-05-31 12:19:59 | DELETED |
+--------------------------------------------------------------------------------------------------+-----------------------------+---------------------+---------+
```

### List active Api in a compartment

```sql
select
  id,
  display_name,
  time_created,
  lifecycle_state as state
from
  oci_apigateway_api
where
  lifecycle_state = 'ACTIVE';
```
```
+--------------------------------------------------------------------------------------------------+-----------------+---------------------+--------+
| id                                                                                               | display_name    | time_created        | state  |
+--------------------------------------------------------------------------------------------------+-----------------+---------------------+--------+
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaarzd76rlohddfi4b2rvyjrhilpdqgz6j2zkovc2bh7jia | turbotsteampipe | 2021-05-27 13:22:37 | ACTIVE |
| ocid1.apigatewayapi.oc1.ap-mumbai-1.amaaaaaa6igdexaaitbtzu76ilnaatrkayscncje522kqyboo4ztg4swhhca | turbotsteampipe | 2021-05-27 13:32:40 | ACTIVE |
+--------------------------------------------------------------------------------------------------+-----------------+---------------------+--------+
```

### List api older than 90 days

```sql
select
  id,
  lifecycle_state,
  time_created
from
  oci_apigateway_api
where
  time_created <= (current_date - interval '90' day)
order by
  time_created;
```
```
+----+-----------------+--------------+
| id | lifecycle_state | time_created |
+----+-----------------+--------------+
+----+-----------------+--------------+
```

### CODE REVIEW

File: docs/tables/oci_apigateway_api.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  id,
+  display_name,
+  time_created,
+  lifecycle_state as state
+from
+  oci_apigateway_api;
+```
+
+
+### List active Api in a compartment
```
Comment (khushboo9024) [inline docs/tables/oci_apigateway_api.md:20]:
```suggestion
### List active apis
```

File: oci-test/tests/oci_apigateway_api/test-get-query.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+select id, display_name
```
Comment (khushboo9024) [inline oci-test/tests/oci_apigateway_api/test-get-query.sql:1]: Please add more columns.

File: oci/service.go [hunk 3]
```diff
@@ -35,7 +36,9 @@ import (
 )
 
 type session struct {
-	TenancyID                      string
+	TenancyID string
```
Comment (khushboo9024) [inline oci/service.go:39]: Indentation is wrong.

File: oci/table_oci_apigateway_api.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+)
+
+//// TABLE DEFINITION
+
+func tableApiGatewayApi(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "oci_apigateway_api",
+		Description: "OCI Apigateway Api",
+		Get: &plugin.GetConfig{
+			KeyColumns: plugin.AnyColumn([]string{"id"}),
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:22]:
```suggestion
				KeyColumns: plugin.SingleColumn("id"),
```

File: oci/table_oci_apigateway_api.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+				Name:        "time_created",
+				Description: "The time this resource was created.",
+				Type:        proto.ColumnType_TIMESTAMP,
+				Transform:   transform.FromField("TimeCreated.Time"),
+			},
+
+			// other columns
+			{
+				Name:        "time_updated",
+				Description: "The OCID of dedicated VM host.",
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:51]:
```suggestion
				Description: "The time this resource was last updated.",
```

File: oci/table_oci_apigateway_api.go [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+			{
+				Name:        "title",
+				Description: ColumnDescriptionTitle,
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("DisplayName"),
+			},
+
+			// Standard OCI columns
+			{
+				Name:        "region",
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:106]: If this is compartment level, then we don't need the region column.

File: oci/table_oci_apigateway_api.go [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+				Hydrate:     getTenantId,
+				Transform:   transform.FromValue(),
+			},
+		},
+	}
+}
+
+//// LIST FUNCTION
+
+func listApigatewayApis(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:130]:
```suggestion
func listApiGatewayApis(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```
Please make changes where ever required.

File: oci/table_oci_apigateway_api.go [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+			pagesLeft = false
+		}
+	}
+
+	return nil, err
+}
+
+//// HYDRATE FUNCTION
+
+func getApigatewayApi(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:168]:
```suggestion
func getApiGatewayApi(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```

File: oci/table_oci_apigateway_api.go [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+	switch item.(type) {
+	case apigateway.Api:
+		return *item.(apigateway.Api).Id
+	case apigateway.ApiSummary:
+		return *item.(apigateway.ApiSummary).Id
+	}
+	return ""
+}
+
+func apigatewayApiFreeformTags(item interface{}) map[string]string {
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:250]:
```suggestion
func apiGatewayApiFreeformTags(item interface{}) map[string]string {
```

File: oci/table_oci_apigateway_api.go [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+	switch item.(type) {
+	case apigateway.Api:
+		return item.(apigateway.Api).FreeformTags
+	case apigateway.ApiSummary:
+		return item.(apigateway.ApiSummary).FreeformTags
+	}
+	return nil
+}
+
+func apigatewayApiDefinedTags(item interface{}) map[string]map[string]interface{} {
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:260]:
```suggestion
func apiGatewayApiDefinedTags(item interface{}) map[string]map[string]interface{} {
```

File: oci/table_oci_apigateway_api.go [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+
+			// other columns
+			{
+				Name:        "time_updated",
+				Description: "The OCID of dedicated VM host.",
+				Type:        proto.ColumnType_TIMESTAMP,
+				Transform:   transform.FromField("TimeUpdated.Time"),
+			},
+			{
+				Name:        "lifecycle_state",
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:56]: `lifecycle_state` should be the third column

File: oci/table_oci_apigateway_api.go [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+				Description: "A message describing the current lifecycleState",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "specification_type",
+				Description: "Type of API Specification file.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "validation_results",
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:71]: `validation_results` should come under JSON fields.

File: oci/table_oci_apigateway_api.go [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+				Transform:   transform.FromField("TimeUpdated.Time"),
+			},
+			{
+				Name:        "lifecycle_state",
+				Description: "The current state of the API.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "lifecycle_details",
+				Description: "A message describing the current lifecycleState",
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:62]:
```suggestion
				Description: "A message describing the current lifecycleState.",
```

File: oci/table_oci_apigateway_api.go [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+	compartment := plugin.GetMatrixItem(ctx)[matrixKeyCompartment].(string)
+
+	// Restrict the api call to only root compartment/ per region
+	if !strings.HasPrefix(compartment, "ocid1.tenancy.oc1") {
+		return nil, nil
+	}
+
+	var id string
+	if h.Item != nil {
+		id = apiId(h.Item)
```
Comment (khushboo9024) [inline oci/table_oci_apigateway_api.go:179]: Is this block required ? we are not call get api from list call.

**Review by khushboo9024 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**